app/jobs/pubsub_job.rb

```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+class PubsubJob < ApplicationJob
+  queue_as(:default)
+  retry_on(StandardError, wait: 5.minutes, attempts: 3)
+
+  self.queue_adapter = :pubsub
+
+  def perform(data)
+    Rails.logger.info("[PubsubJob] Perform Now: #{data}")
+
+    raise(StandardError) if rand < 0.2 # fail 20% of the time
+
+    Pubsub.publish!(data)
+  end
+
+  before_enqueue do |job|
+    # Before Enqueueing:
+    # If it's already the 3th time, meaning the first,
+    # and second try failed. Then use :morgue as queue
+    job.queue_name = :morgue if job.executions >= 2 && job.queue_name != :morgue
+  end
```

Comment:
This should happen after it fails, when Rails bubbles up the error - it won't get here.

Reply:
I tested this locally and it's able to change the queue name for the third attempt when the first two attempt fails.

I understand the evaluation is done but I would appreciate if you could explain further.

Reply:
If you use retry_on with an upper limit, once retries are exhausted it will error out. Putting the logic for retries here bypasses ActiveJob's interfaces for error handling.

Reply:
I see. Thanks for explaining. Will have a look at an alternative approach.